src/LayerEditor/ui/panels/regions_panel.py

```diff
@@ -173,98 +162,104 @@ def __init__(self, layer_heads, parent=None):
         :return:
         """
-        # Update number of tabs and their layers.
-        self.layer_heads.reinit_regions_map()
-        i_tab = 0
-        names = self.layer_heads.layer_names
-        self.layer_id_to_idx = {}
-        for layer_id, layer_name in names:
-            self.layer_id_to_idx[layer_id] = i_tab
-            if i_tab >= self.count():
-                tab_widget = RegionLayerTab(self.layer_heads, i_tab, self)
-                self.tabs.append(tab_widget)
+        with nosignal(self):
+            while self.count() > 0:
+                self.removeItem(0)
+            names = self.layer_heads.layer_names
+            for idx, layer_name in enumerate(names):
+                tab_widget = RegionLayerTab(self.layer_heads, self.layer_heads.block.layers[idx], self)
                 self.addItem(tab_widget, "")
-            self.tabs[i_tab].reinit(layer_id, layer_name)
-            i_tab += 1
-        while i_tab < self.count():
-            self.removeItem(i_tab)
-            self.tabs.pop(-1)
-            i_tab += 1
-        # Update content.
-        self.setCurrentIndex(self.layer_id_to_idx[self.layer_heads.current_layer_id])
+                tab_widget.reinit(layer_name)
+            # Update content.
+        self.setCurrentIndex(self.layer_heads.current_layer_idx)
 
     def _update_tab_head(self, tab):
         """
         Update tab head, called by tab itself after change of region.
-        :param i:
+        :param tab:
         :return:
         """
         color = tab.region_color
-        item_idx = self.layer_id_to_idx[tab.layer_id]
+        item_idx = self.layer_heads.block.layers.index(tab.layer)
         pixmap = QtGui.QPixmap(16, 16)
         pixmap.fill(color)
         iconPix = QtGui.QIcon(pixmap)
         self.setItemIcon(item_idx, iconPix)
-        self.setItemText(item_idx, "{} ({})".format(tab._layer_name, tab.region.name))
+        self.setItemText(item_idx, "{} ({})".format(tab._layer_name, tab.curr_region.name))
 
     def add_region(self):
         """
         Add new region to all combo and select it in current tab.
         Handle combo changed signal in current tab.
         """
-        dialog = AddRegionDlg(self.layer_heads.max_selected_dim, self.layer_heads.region_names, self)
+        dialog = AddRegionDlg(self.layer_heads.max_selected_dim(), self.layer_heads.region_names, self)
         dialog_result = dialog.exec_()
         if dialog_result == QtWidgets.QDialog.Accepted:
             name = dialog.region_name.text()
             dim = dialog.region_dim.currentData()
-            color = dialog.get_some_color(self.layer_heads.regions._get_available_reg_id() - 1).name()
-            region = self.layer_heads.add_region(name, dim, color)
-            self.current_tab._combo_set_region(region.reg_id)
-            self.update_tabs()
-            self.color_changed.emit()
+            region = self.layer_heads.add_region(name, dim)
```

Comment:
What about colors?

Reply:
Updating was unified so every update of scene also updates colors.